horizon/enforcer/utils/statistics_utils.py

```diff
@@ -1,47 +1,77 @@
         self._interval_seconds = interval_seconds
+        self._failures_threshold_percentage = failures_threshold_percentage
+        self._had_failure = False
 
     async def restarter_task(self) -> None:
         while True:
             message = await self._messages.get()
             try:
                 async with self._lock:
+                    logger.debug("Statistics message: {message}", message=message)
                     self._requests += 1
                     if message is False:
                         self._failures += 1
             finally:
                 self._messages.task_done()
-            await asyncio.sleep(0.5)
+
+    async def reset_stats(self) -> None:
+        async with self._lock:
+            logger.info("Interval task: requests={requests}, failures={failures}",
+                        requests=self._requests, failures=self._failures)
+            self._requests = 0
+            self._failures = 0
 
     async def interval_task(self) -> None:
         while True:
             await asyncio.sleep(self._interval_seconds)
-            async with self._lock:
-                self._requests = 0
-                self._failures = 0
+            await self.reset_stats()
 
-    def run(self) -> None:
-        self._restarter_task = asyncio.create_task(self.restarter_task())
-        self._interval_task = asyncio.create_task(self._interval_task())
+    async def run(self) -> None:
+        logger.debug("Starting statistics manager")
+        if self._restarter_task is None:
+            self._restarter_task = asyncio.create_task(self.restarter_task())
+        if self._interval_task is None:
+            self._interval_task = asyncio.create_task(self.interval_task())
+
+    async def stop(self) -> None:
+        logger.debug("Stopping statistics manager")
+        await self._messages.join()
+        if self._restarter_task is not None:
+            self._restarter_task.cancel()
+            self._restarter_task = None
+        if self._interval_task is not None:
+            self._interval_task.cancel()
+            self._interval_task = None
 
     def report_success(self) -> None:
+        logger.info("Reporting success")
         self._messages.put_nowait(True)
 
     def report_failure(self) -> None:
+        logger.info("Reporting failure")
         self._messages.put_nowait(False)
 
     async def current_rate(self) -> float:
-        async with self._lock:
-            current_requests = float(self._requests)
-            current_failures = float(self._failures)
+        current_requests, current_failures = float(self._requests), float(self._failures)
+        if current_requests == 0:
+            return 0.0
         return current_failures / current_requests
+
+    async def status(self) -> bool:
+        rate = await self.current_rate()
+        if rate > self._failures_threshold_percentage:
+            self._had_failure = True
+        return self._had_failure
```

Comment:
not sure if this should be separated into a separate check or not.
i.e:
- currently having failures
- had failures (at least once)

not a blocker, just wondering.